src/controllers/hid/hidiothread.cpp

```diff
@@ -0,0 +1,234 @@
+#include "controllers/hid/hidiothread.h"
+
+#include <hidapi.h>
+
+#include <QTimer>
+
+#include "controllers/defs_controllers.h"
+#include "controllers/hid/legacyhidcontrollermappingfilehandler.h"
+#include "moc_hidiothread.cpp"
+#include "util/string.h"
+#include "util/time.h"
+#include "util/trace.h"
+
+namespace {
+constexpr int kReportIdSize = 1;
+constexpr int kMaxHidErrorMessageSize = 512;
+QString loggingCategoryPrefix(const QString& deviceName) {
+    return QStringLiteral("controller.") +
+            RuntimeLoggingCategory::removeInvalidCharsFromCategory(deviceName.toLower());
+}
+} // namespace
+
+HidIoThread::HidIoThread(
+        hid_device* device, const mixxx::hid::DeviceInfo&& deviceInfo)
+        : QThread(),
+          m_pollingBufferIndex(0),
+          // Defining RuntimeLoggingCategories locally in this thread improves runtime performance significiant
+          m_logBase(loggingCategoryPrefix(deviceInfo.formatName())),
+          m_logInput(loggingCategoryPrefix(deviceInfo.formatName()) +
+                  QStringLiteral(".input")),
+          m_logOutput(loggingCategoryPrefix(deviceInfo.formatName()) +
+                  QStringLiteral(".output")),
+          m_pHidDevice(device),
+          m_deviceInfo(std::move(deviceInfo)),
+          m_HidDeviceMutex(QT_RECURSIVE_MUTEX_INIT),
+          mPollTimerId(0) {
+    // This isn't strictly necessary but is good practice.
+    for (int i = 0; i < kNumBuffers; i++) {
+        memset(m_pPollData[i], 0, kBufferSize);
+    }
+    m_lastPollSize = 0;
+}
+
+void HidIoThread::timerEvent(QTimerEvent* event) {
+    Q_UNUSED(event);
+    poll();
+}
+
+void HidIoThread::startPollTimer() {
+    if (mPollTimerId == 0) {
+        mPollTimerId = startTimer(1, Qt::PreciseTimer);
+    }
+}
+
+void HidIoThread::stopPollTimer() {
+    qCInfo(m_logBase) << "Stop HidIoThread polling"
+                      << m_deviceInfo.formatName();
+    killTimer(mPollTimerId);
+}
+
+void HidIoThread::poll() {
+    Trace hidRead("HidIoThread poll");
+    auto lock = lockMutex(&m_HidDeviceMutex);
+    // This loop risks becoming a high priority endless loop in case processing
+    // the mapping JS code takes longer than the controller polling rate.
+    // This could stall other low priority tasks.
+    // There is no safety net for this because it has not been demonstrated to be
+    // a problem in practice.
+    while (true) {
```

Comment:
You should also check for timer shutdown here as another exit condition

Reply:
I added a condition, but as long the loop runs, nothing can shutdown the timer. Both runs in the same thread.
This is not different to now.

Reply:
Now there is a public atomic int, which can be set from outside, while this loop is running. This ensures an immediate stop of reading InputReports.